pkg/hive/manager.go

```diff
@@ -4,36 +4,112 @@ package hive
 // Licensed under the Apache License 2.0.
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
 	hiveclient "github.com/openshift/hive/pkg/client/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/retry"
 )
 
-type Interface interface {
+type ClusterManager interface {
+	// We need to keep cluster SP updated with Hive, so we not only need to be able to create,
+	// but also be able to update resources.
+	// See relevant work item: https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/14895480
+	// TODO: Replace Register with CreateOrUpdate and remove the comment above
+	Register(ctx context.Context, workloadCluster *WorkloadCluster) (*hivev1.ClusterDeployment, error)
+	IsConnected(ctx context.Context, namespace string) (bool, string, error)
+}
+
+// WorkloadCluster represents all data in hive pertaining to a single ARO cluster
+type WorkloadCluster struct {
+	SubscriptionID    string `json:"subscription,omitempty"`
+	ClusterName       string `json:"name,omitempty"`
+	ResourceGroupName string `json:"resourceGroup,omitempty"`
+	Location          string `json:"location,omitempty"`
+	InfraID           string `json:"infraId,omitempty"`
+	ClusterID         string `json:"clusterID,omitempty"`
+	KubeConfig        string `json:"kubeconfig,omitempty"`
+	ServicePrincipal  string `json:"serviceprincipal,omitempty"`
 }
 
-type client struct {
+type clusterManager struct {
 	hiveClientset *hiveclient.Clientset
 	kubernetescli *kubernetes.Clientset
 }
 
-func New(hiveClientset *hiveclient.Clientset, kubernetescli *kubernetes.Clientset) Interface {
-	return &client{
+func NewClusterManagerFromConfig(restConfig *rest.Config) (ClusterManager, error) {
+	hiveclientset, err := hiveclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	kubernetescli, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return newClusterManager(hiveclientset, kubernetescli), nil
+}
+
+func newClusterManager(hiveClientset *hiveclient.Clientset, kubernetescli *kubernetes.Clientset) ClusterManager {
+	return &clusterManager{
 		hiveClientset: hiveClientset,
 		kubernetescli: kubernetescli,
 	}
 }
 
-func NewFromConfig(restConfig *rest.Config) (Interface, error) {
-	hiveclientset, err := hiveclient.NewForConfig(restConfig)
+func (hr *clusterManager) Register(ctx context.Context, workloadCluster *WorkloadCluster) (*hivev1.ClusterDeployment, error) {
+	var namespace string
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		namespace = "aro-" + uuid.Must(uuid.NewV4()).String()
```

Comment:
I think we should reuse the existing UUID that we have for the cluster -- the document ID. There's a few reasons for this:

1. It is common to the cluster, not just our database. In generating the installconfig, we set the cluster UUID as the document ID.
2. It means we can remove the HiveProfile struct. We may need it for other things later, but we at least don't need to list the namespace.
3. At runtime, since it is common to the cluster, it can assist us in working backwards from a cluster in production to the Hive namespace, without having to do a second lookup from the RP.

Reply:
@hawkowl this solves (I think?) the issue of uniqueness as we can assume that DB ids are unique and we don't have to retry on conflict here. And, as you said, we will be able to work from cluster to hive.

However very common SRE/tooling use case will be working from IcM to Hive (troubleshooting failed provisioning/deprovisioning/updates). In IcM we will have a cluster ID and a region. With that we can get [`OpenShiftCluster`](https://github.com/Azure/ARO-RP/blob/8f0a38c1919e54a9cf3931a361ca98f26534f858/pkg/api/openshiftcluster.go#L11-L23), but not [`OpenShiftClusterDocument`](https://github.com/Azure/ARO-RP/blob/8f0a38c1919e54a9cf3931a361ca98f26534f858/pkg/api/openshiftclusterdocument.go#L20-L48) which contains an the ID you suggest to use.

So we won't have a way of figuring out namespace on the AKS/Hive cluster via Geneva action for a specific cluster in many cases (failed provisioning or when API server is down).

Reply:
Won't our basic checks automation be able to spit out that ID for us?

Reply:
@hawkowl no. Basic checks get cluster endpoint and we marshall `doc.OpenShiftCluster`, not `doc` so `doc.ID` is not included in the response.